HttpSender fails because of NetworkOnMainThreadException

Hi

Currently using
5.3.0-rc01

but have not had reports for some time now... (since migrating to 5)

I have a custom activity that extends BaseCrashReportDialog.

In a click listener on a button I call

`sendCrash("", "");`

the reports never arrive.

This is what I get on the :acra process

`
2019-02-09 22:19:05.280 31168-31168/? D/acra: acra
2019-02-09 22:19:06.050 31168-31168/com.test.app:acra I/ACRA: Sending report /data/user/0/com.test.app/app_ACRA-approved/2019-02-09T21:09:20.317+00:00.stacktrace
2019-02-09 22:19:06.131 31168-31168/com.test.app:acra D/NetworkSecurityConfig: No Network Security Config specified, using platform default
2019-02-09 22:19:06.148 31168-31168/com.test.app:acra E/ACRA: Failed to send crash report for /data/user/0/com.test.app/app_ACRA-approved/2019-02-09T21:09:20.317+00:00.stacktrace
    org.acra.sender.ReportSenderException: Policy marked this task as incomplete. ACRA will try to send this report again.
        at org.acra.sender.ReportDistributor.sendCrashReport(ReportDistributor.java:123)
        at org.acra.sender.ReportDistributor.distribute(ReportDistributor.java:76)
        at org.acra.sender.SendingConductor.sendReports(SendingConductor.java:63)
        at org.acra.sender.JobSenderService.onStartJob(JobSenderService.java:23)
        at android.app.job.JobService$1.onStartJob(JobService.java:62)
        at android.app.job.JobServiceEngine$JobHandler.handleMessage(JobServiceEngine.java:108)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:193)
        at android.app.ActivityThread.main(ActivityThread.java:6669)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
     Caused by: org.acra.sender.ReportSenderException: Error while sending JSON report via Http PUT
        at org.acra.sender.HttpSender.send(HttpSender.java:137)
        at org.acra.sender.ReportDistributor.sendCrashReport(ReportDistributor.java:111)
        at org.acra.sender.ReportDistributor.distribute(ReportDistributor.java:76) 
        at org.acra.sender.SendingConductor.sendReports(SendingConductor.java:63) 
        at org.acra.sender.JobSenderService.onStartJob(JobSenderService.java:23) 
        at android.app.job.JobService$1.onStartJob(JobService.java:62) 
        at android.app.job.JobServiceEngine$JobHandler.handleMessage(JobServiceEngine.java:108) 
        at android.os.Handler.dispatchMessage(Handler.java:106) 
        at android.os.Looper.loop(Looper.java:193) 
        at android.app.ActivityThread.main(ActivityThread.java:6669) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858) 
     **Caused by: android.os.NetworkOnMainThreadException
        at android.os.StrictMode$AndroidBlockGuardPolicy.onNetwork(StrictMode.java:1513)**
        at java.net.Inet6AddressImpl.lookupHostByName(Inet6AddressImpl.java:117)
        at java.net.Inet6AddressImpl.lookupAllHostAddr(Inet6AddressImpl.java:105)
        at java.net.InetAddress.getAllByName(InetAddress.java:1154)
        at com.android.okhttp.Dns$1.lookup(Dns.java:39)
        at com.android.okhttp.internal.http.RouteSelector.resetNextInetSocketAddress(RouteSelector.java:175)
        at com.android.okhttp.internal.http.RouteSelector.nextProxy(RouteSelector.java:141)
        at com.android.okhttp.internal.http.RouteSelector.next(RouteSelector.java:83)
        at com.android.okhttp.internal.http.StreamAllocation.findConnection(StreamAllocation.java:174)
        at com.android.okhttp.internal.http.StreamAllocation.findHealthyConnection(StreamAllocation.java:126)
        at com.android.okhttp.internal.http.StreamAllocation.newStream(StreamAllocation.java:95)
        at com.android.okhttp.internal.http.HttpEngine.connect(HttpEngine.java:281)
        at com.android.okhttp.internal.http.HttpEngine.sendRequest(HttpEngine.java:224)
        at com.android.okhttp.internal.huc.HttpURLConnectionImpl.execute(HttpURLConnectionImpl.java:461)
        at com.android.okhttp.internal.huc.HttpURLConnectionImpl.connect(HttpURLConnectionImpl.java:127)
        at com.android.okhttp.internal.huc.DelegatingHttpsURLConnection.connect(DelegatingHttpsURLConnection.java:89)
        at com.android.okhttp.internal.huc.HttpsURLConnectionImpl.connect(HttpsURLConnectionImpl.java:26)
        at org.acra.http.BaseHttpRequest.writeContent(BaseHttpRequest.java:184)
        at org.acra.http.BaseHttpRequest.send(BaseHttpRequest.java:106)
        at org.acra.sender.HttpSender.sendWithoutAttachments(HttpSender.java:166)
        at org.acra.sender.HttpSender.sendHttpRequests(HttpSender.java:154)
        at org.acra.sender.HttpSender.send(HttpSender.java:132)
        at org.acra.sender.ReportDistributor.sendCrashReport(ReportDistributor.java:111) 
        at org.acra.sender.ReportDistributor.distribute(ReportDistributor.java:76) 
        at org.acra.sender.SendingConductor.sendReports(SendingConductor.java:63) 
        at org.acra.sender.JobSenderService.onStartJob(JobSenderService.java:23) 
        at android.app.job.JobService$1.onStartJob(JobService.java:62) 
        at android.app.job.JobServiceEngine$JobHandler.handleMessage(JobServiceEngine.java:108) 
        at android.os.Handler.dispatchMessage(Handler.java:106) 
        at android.os.Looper.loop(Looper.java:193) 
        at android.app.ActivityThread.main(ActivityThread.java:6669) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858) 
`

Would appreciate any help...

Thanks

--- a/acra-core/src/main/java/org/acra/sender/JobSenderService.java
+++ b/acra-core/src/main/java/org/acra/sender/JobSenderService.java
@@ -20,9 +20,12 @@ public class JobSenderService extends JobService {
         CoreConfiguration config = IOUtils.deserialize(CoreConfiguration.class, extras.getString(LegacySenderService.EXTRA_ACRA_CONFIG));
         boolean onlySilent = extras.getBoolean(LegacySenderService.EXTRA_ONLY_SEND_SILENT_REPORTS);
         if (config != null) {
-            new SendingConductor(this, config).sendReports(onlySilent);
+            new Thread(() -> {
+                new SendingConductor(this, config).sendReports(onlySilent);
+                jobFinished(params, false);
+            }).start();
         }
-        return false;
+        return true;
     }
 
     @Override

--- a/acra-core/src/main/java/org/acra/sender/LegacySenderService.java
+++ b/acra-core/src/main/java/org/acra/sender/LegacySenderService.java
@@ -34,7 +34,10 @@ public class LegacySenderService extends Service {
         if (intent.hasExtra(EXTRA_ACRA_CONFIG)) {
             final boolean onlySendSilentReports = intent.getBooleanExtra(EXTRA_ONLY_SEND_SILENT_REPORTS, false);
             final CoreConfiguration config = (CoreConfiguration) intent.getSerializableExtra(EXTRA_ACRA_CONFIG);
-            new SendingConductor(this, config).sendReports(onlySendSilentReports);
+            new Thread(() -> {
+                new SendingConductor(this, config).sendReports(onlySendSilentReports);
+                stopSelf();
+            }).start();
         } else {
             if (ACRA.DEV_LOGGING) ACRA.log.d(LOG_TAG, "SenderService was started but no valid intent was delivered, will now quit");
         }

